satellite/gracefulexit/endpoint_test.go

```diff
@@ -378,13 +378,156 @@ func TestInvalidStorageNodeSignature(t *testing.T) {
+		progress, err := satellite.DB.GracefulExit().GetProgress(ctx, exitingNode.ID())
+		require.NoError(t, err)
+
+		require.Equal(t, int64(0), progress.PiecesTransferred)
+		require.Equal(t, int64(1), progress.PiecesFailed)
+	})
+
+}
+
+func TestExitDisqualifiedNodeFailEventually(t *testing.T) {
+	testTransfers(t, numObjects, func(ctx *testcontext.Context, nodeFullIDs map[storj.NodeID]*identity.FullIdentity, satellite *testplanet.SatelliteSystem, processClient exitProcessClient, exitingNode *storagenode.Peer, numPieces int) {
+		disqualifyNode(t, ctx, satellite, exitingNode.ID())
+		var pieceID storj.PieceID
+		failedCount := 0
+		deletedCount := 0
+		for {
+			response, err := processClient.Recv()
+			if errs.Is(err, io.EOF) {
+				// Done
+				break
+			}
+			if deletedCount >= numPieces {
+				// when a disqualified node has finished transfer all pieces, it should receive an error
+				require.True(t, errs2.IsRPC(err, rpcstatus.PermissionDenied))
+				break
+			} else {
+				require.NoError(t, err)
+			}
+
+			switch m := response.GetMessage().(type) {
+			case *pb.SatelliteMessage_TransferPiece:
+				require.NotNil(t, m)
+
+				// pick the first one to fail
+				if pieceID.IsZero() {
+					pieceID = m.TransferPiece.OriginalPieceId
+				}
+
+				if failedCount > 0 || pieceID != m.TransferPiece.OriginalPieceId {
+
+					pieceReader, err := exitingNode.Storage2.Store.Reader(ctx, satellite.ID(), m.TransferPiece.OriginalPieceId)
+					require.NoError(t, err)
+
+					header, err := pieceReader.GetPieceHeader()
+					require.NoError(t, err)
+
+					orderLimit := header.OrderLimit
+					originalPieceHash := &pb.PieceHash{
+						PieceId:   orderLimit.PieceId,
+						Hash:      header.GetHash(),
+						PieceSize: pieceReader.Size(),
+						Timestamp: header.GetCreationTime(),
+						Signature: header.GetSignature(),
+					}
+
+					newPieceHash := &pb.PieceHash{
+						PieceId:   m.TransferPiece.AddressedOrderLimit.Limit.PieceId,
+						Hash:      originalPieceHash.Hash,
+						PieceSize: originalPieceHash.PieceSize,
+						Timestamp: time.Now(),
+					}
+
+					receivingNodeID := nodeFullIDs[m.TransferPiece.AddressedOrderLimit.Limit.StorageNodeId]
+					require.NotNil(t, receivingNodeID)
+					signer := signing.SignerFromFullIdentity(receivingNodeID)
+
+					signedNewPieceHash, err := signing.SignPieceHash(ctx, signer, newPieceHash)
+					require.NoError(t, err)
+
+					success := &pb.StorageNodeMessage{
+						Message: &pb.StorageNodeMessage_Succeeded{
+							Succeeded: &pb.TransferSucceeded{
+								OriginalPieceId:      m.TransferPiece.OriginalPieceId,
+								OriginalPieceHash:    originalPieceHash,
+								OriginalOrderLimit:   &orderLimit,
+								ReplacementPieceHash: signedNewPieceHash,
+							},
+						},
+					}
+					err = processClient.Send(success)
+					require.NoError(t, err)
+				} else {
+					failedCount++
+					failed := &pb.StorageNodeMessage{
```

Comment:
I'm not sure I understand why we fail any piece transfers. This test is ensuring that if the storagenode is disqualified when a graceful exit is already in progress, that eventually graceful exit fails because of the disqualification. Because of that, I think we should only send valid transfer success messages, and still expect graceful exit failure + permission denied at the end.